dvc/repo/metrics/show.py

```diff
@@ -123,20 +116,47 @@ def _collect_metrics(self, path, recursive, typ, xpath, branch):
     return res
 
 
+def _read_metrics_filesystem(self, path, typ, xpath):
+    if not os.path.exists(path):
+        return None
+    with open(path, "r") as fd:
+        return _read_metric(self, fd, typ=typ, xpath=xpath)
+
+
 def _read_metrics(self, metrics, branch):
     res = {}
     for out, typ, xpath in metrics:
-        assert out.scheme == "local"
-        if out.use_cache:
-            path = self.cache.local.get(out.checksum)
-        else:
-            path = out.path
-
-        metric = _read_metric(path, typ=typ, xpath=xpath, branch=branch)
-        if not metric:
-            continue
-
-        res[out.rel_path] = metric
+        try:
+            assert out.scheme == "local"
+            if out.use_cache:
+                metric = _read_metrics_filesystem(
+                    self,
+                    self.cache.local.get(out.checksum),
+                    typ=typ,
+                    xpath=xpath,
+                )
+            elif not branch:
+                metric = _read_metrics_filesystem(
+                    self, out.path, typ=typ, xpath=xpath
+                )
+            else:
+                fd = self.scm.open(out.path, branch)
+                metric = _read_metric(self, fd, typ=typ, xpath=xpath)
+
+            if not metric:
+                continue
+
+            res[out.rel_path] = metric
+
+        # Json path library has to be replaced or wrapped in
+        # order to fix this too broad except clause.
+        except Exception:
```

Comment:
moving it here has two problems - catching even more potential problems that should not be caught (especially considering that it's `except Exception`). Second, overloading more general method with details that belong to the `_read_metric`. I would find a way to pass the context into it to print the message. I think even Code Climate will complain about complexity of this method, btw.

Reply:
I think we should go with our own exception and move it to the `_read_metric_<format>` functions.

Reply:
you are right (we had a comment to fix that)! but before we jump into this, let's not make the scope of this catch even wider